crypto/pkcs7/pkcs7.c

```diff
@@ -191,3 +194,415 @@ int pkcs7_add_signed_data(CBB *out,
 
   return CBB_flush(out);
 }
+
+int PKCS7_set_type(PKCS7 *p7, int type) {
+    if (p7 == NULL) {
+        OPENSSL_PUT_ERROR(PKCS7, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    ASN1_OBJECT *obj = OBJ_nid2obj(type);
+    if (obj == NULL) {
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_UNSUPPORTED_CONTENT_TYPE);
+        return 0;
+    }
+
+    switch (type) {
+    case NID_pkcs7_signed:
+        p7->type = obj;
+        p7->d.sign = PKCS7_SIGNED_new();
```

Comment:
Prior to the `p7->d.sign = PKCS7_SIGNED_new()` assignment, do we need to check whether a value is already assigned and free it if so? (Likewise for the other cases.)

Reply:
yes, nice catch. i'll update this and the tests.